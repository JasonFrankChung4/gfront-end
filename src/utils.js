/*global ASSET_PATH */

import get from 'lodash.get'
import intersection from 'lodash.intersection'
import { run, startsWith } from '@code.gov/cautious'

export const falses = [undefined, null, 'null', 'None', 'Null', 'NULL', '', 'False', 'false']

export function isFalse(input) {
  return falses.includes(input)
}

export function isSet(input) {
  return input && typeof input === 'object' && input.has && input.add
}

export function adjustAssetPath(thing) {
  const pattern = /.?\/?assets\//
  const newAssetPath = ASSET_PATH + 'assets/'
  if (startsWith(thing, './assets') || startsWith(thing, '/assets/') || startsWith(thing, 'assets/')) {
    return thing.replace(pattern, newAssetPath)
  } else if (typeof value === 'object') {
    for (let key in thing) {
      const subvalue = thing[key]
      if (typeof subvalue === 'string') {
        thing[key] = adjustAssetPath(subvalue)
      }
    }
    return thing
  } else {
    return thing
  }
}

export function getConfigValue(siteConfig, path) {
  if (siteConfig) {
    let value = get(siteConfig, path)
    if (typeof value === 'string') {
      try {
        value = adjustAssetPath(value)
      } catch (error) {
        console.error("couldn't check if the following started with assets", [value])
      }
    } else if (Array.isArray(value)) {
      value = value.map(item => {
        if (typeof item === 'object') {
          for (let subkey in item) {
            let subvalue = item[subkey]
            if (typeof subvalue === 'string') {
              item[subkey] = adjustAssetPath(subvalue)
            }
          }
        }
        return item
      });
    }
    if (!value) {
      console.warn(`We weren't able to find the value for ${path} in your code-gov-config.json file.
      You can consult examples under the folder found at config/site/examples
      or consult our developer documentation here: https://developers.code.gov/configure.html`)
    }
    return value
  }
  else {
    return null
  }
}

export function normalize(input) {
  return trim(lower(input))
}

export function onEachItem(input, func) {
  if (Array.isArray(input)) {
    return input.map(item => run(item, func))
  } else if (isSet(input)) {
    return Set(Array.from(input).map(item => run(item, func)))
  } else {
    return run(input, func)
  }
}

export function lower(input) {
  return onEachItem(input, 'toLowerCase')
}

export function upper(input) {
  return onEachItem(input, 'toUpperCase')
}

export function trim(input) {
  return onEachItem(input, 'trim')
}

/* runs when each page component is loaded */
export function refreshView() {
  window.scrollTo(0, 0)
  document.activeElement.blur()
}


export function scrollToTopOfResults() {
  const headerHeight = document.querySelector("header.main").clientHeight
  const bannerHeight = document.querySelector(".banner").clientHeight
  const navHeight = document.querySelector("header.main nav").clientHeight
  const scrollDepth = headerHeight + bannerHeight - navHeight
  window.scrollTo(0, scrollDepth)
}

/* gets a set of values given a path */
export function getSet(items, path) {
  const results = new Set()
  if (Array.isArray(items) && typeof path === 'string') {
    items.forEach(item => {
      const value = get(item, path)
      if (value && value !== '') {
        if (Array.isArray(value)) {
          value.forEach(subitem => {
            if (isFalse(subitem) === false) {
              results.add(subitem)
            }
          })
        } else {
          if (isFalse(value) === false) {
            results.add(value)
          }
        }
      }
    })
  }
  return results
}

export function getLowerSet(items, path) {
  const results = new Set()
  if (Array.isArray(items) && typeof path === 'string') {
    items.forEach(item => {
      const value = get(item, path)
      if (value && value !== '') {
        if (Array.isArray(value)) {
          value.forEach(subitem => {
            if (isFalse(subitem) === false) {
              results.add(subitem.toLowerCase())
            }
          })
        } else {
          if(isFalse(value) === false) {
            results.add(value.toLowerCase())
          }
        }
      }
    })
  }
  return results
}

export function getFilterData(key, path, currentSearchResults, filters) {
  if (currentSearchResults && filters && filters[key]) {
    const names = getLowerSet(currentSearchResults.repos, path)
    return filters[key].filter(({ name, value }) => {
      return names.has(normalize(name)) || names.has(normalize(value))
    })
  }
}

export function hasLicense(repo) {
  return repo.permissions && Array.isArray(repo.permissions.licenses) && repo.permissions.licenses.length > 0
}

export function clean(string) {
  try {
    return string.toLowerCase().replace(/-/g,' ').trim()
  } catch (error) {
    return string
  }
}

export function sortByBestMatch(a, b) {
  return Number(a.searchScore) > Number(b.searchScore) ? -1 : 1
}

export function sortByName(a, b) {
  return clean(a.name) > clean(b.name) ? 1 : -1
}

export function getDate(repo, path) {

  if (repo.ghUpdatedAt) {
    return new Date(a.ghUpdatedAt).getTime()
  } else if (repo.date && repo.date.lastModified) {
    return new Date(a.date.lastModified).getTime()
  } else {
    return
  }
}

export function getTime(date) {
  try {
    if (date) {
      return new Date(date).getTime()
    } else {
      return -10e10
    }
  } catch (error) {
    return -10e10
  }
}

export function sortByDate(a, b) {
  try {
    const aDate = get(a, 'date.lastModified') || get(a, 'ghUpdatedAt')
    const bDate = get(b, 'date.lastModified') || get(b, 'ghUpdatedAt')
    const aTime = getTime(aDate)
    const bTime = getTime(bDate)
    return Math.sign(bTime - aTime) || 0;
  } catch (error) {
    return 0
  }
}

export function sortByDataQuality(a, b) {
  return Number(a.score) > Number(b.score) ? -1 : 1
}