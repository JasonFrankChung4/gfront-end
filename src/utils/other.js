/* global ASSET_PATH */
/* global URLSearchParams */

import get from 'lodash.get'
import intersection from 'lodash.intersection'
import { lower, run, startsWith, trim } from '@code.gov/cautious'
import siteConfig from '../../config/site/site.json'

export const falses = [undefined, null, 'null', 'None', 'Null', 'NULL', '', 'False', 'false']

export function isFalse(input) {
  return falses.includes(input)
}


export function adjustAssetPath(thing) {
  const pattern = /.?\/?assets\//
  const newAssetPath = ASSET_PATH + 'assets/'
  if (startsWith(thing, './assets') || startsWith(thing, '/assets/') || startsWith(thing, 'assets/')) {
    return thing.replace(pattern, newAssetPath)
  } else if (typeof value === 'object') {
    for (let key in thing) {
      const subvalue = thing[key]
      if (typeof subvalue === 'string') {
        thing[key] = adjustAssetPath(subvalue)
      }
    }
    return thing
  } else {
    return thing
  }
}

export function getConfigValue(path) {
  if (siteConfig) {
    let value = get(siteConfig, path)
    if (typeof value === 'string') {
      try {
        value = adjustAssetPath(value)
      } catch (error) {
        console.error("couldn't check if the following started with assets", [value])
      }
    } else if (Array.isArray(value)) {
      value = value.map(item => {
        if (typeof item === 'object') {
          for (let subkey in item) {
            let subvalue = item[subkey]
            if (typeof subvalue === 'string') {
              item[subkey] = adjustAssetPath(subvalue)
            }
          }
        }
        return item
      });
    }
    if (!value) {
      console.warn(`We weren't able to find the value for ${path} in your code-gov-config.json file.
      You can consult examples under the folder found at config/site/examples
      or consult our developer documentation here: https://developers.code.gov/configure.html`)
    }
    return value
  }
  else {
    return null
  }
}

export function normalize(input) {
  return trim(lower(input))
}

/* runs when each page component is loaded */
export function refreshView() {
  window.scrollTo(0, 0)
  document.activeElement.blur()
}


export function scrollToTopOfResults() {
  const headerHeight = document.querySelector("header.main").clientHeight
  const bannerHeight = document.querySelector(".banner").clientHeight
  const navHeight = document.querySelector("header.main nav").clientHeight
  const scrollDepth = headerHeight + bannerHeight - navHeight
  window.scrollTo(0, scrollDepth)
}

/* gets a set of values given a path */
export function getSet(items, path) {
  const results = new Set()
  if (Array.isArray(items) && typeof path === 'string') {
    items.forEach(item => {
      const value = get(item, path)
      if (value && value !== '') {
        if (Array.isArray(value)) {
          value.forEach(subitem => {
            if (isFalse(subitem) === false) {
              results.add(subitem)
            }
          })
        } else {
          if (isFalse(value) === false) {
            results.add(value)
          }
        }
      }
    })
  }
  return results
}

export function getLowerSet(items, path) {
  const results = new Set()
  if (Array.isArray(items) && typeof path === 'string') {
    items.forEach(item => {
      const value = get(item, path)
      if (value && value !== '') {
        if (Array.isArray(value)) {
          value.forEach(subitem => {
            if (isFalse(subitem) === false) {
              results.add(subitem.toLowerCase())
            }
          })
        } else {
          if(isFalse(value) === false) {
            results.add(value.toLowerCase())
          }
        }
      }
    })
  }
  return results
}

export function getFilterData(key, path, currentSearchResults, filters) {
  if (currentSearchResults && filters && filters[key]) {
    const names = getLowerSet(currentSearchResults.repos, path)
    return filters[key].filter(({ name, value }) => {
      return names.has(normalize(name)) || names.has(normalize(value))
    })
  }
}

export function hasLicense(repo) {
  return repo.permissions && Array.isArray(repo.permissions.licenses) && repo.permissions.licenses.length > 0
}


