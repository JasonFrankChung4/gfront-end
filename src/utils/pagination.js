import { last, range } from '@code.gov/cautious'

export function getPageInfo(params) {
  const { count, pagesize } = params
  const pagecount = Math.ceil(count / pagesize)
  const pageIndexes = range(pagecount)
    .map(n => n + 1) // convert from starting at 0 to 1
  const pageCount = pageIndexes.length

  return { pageIndexes, pageCount }
}

export function getPaginationNavInfo(params) {
  const { pageIndexes, page } = params
  const ultimate = last(pageIndexes)
  const left = page - 1
  const right = page + 1

  return { ultimate, left, right }
}
