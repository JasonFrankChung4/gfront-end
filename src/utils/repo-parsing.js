import get from 'lodash.get'
import { filter } from '@code.gov/cautious'

export function getDisplayTextForUsageType(repo) {
  let text = ''
  const usageType = get(repo, 'permissions.usageType')
  if (usageType === 'openSource') {
    text = 'Open Source'
  } else if (usageType === 'governmentWideReuse') {
    text = 'Gov-wide Reuse'
  } else if (usageType) {
    text = usageType
  } else {
    text = 'unknown'
  }
  return text
}

export function getLicenseName(repo) {
  return get(repo, 'permissions.licenses[0].name')
}

export function getLaborHours(repo) {
  try {
    const laborHours =  Number(get(repo, 'contact.laborHours'))
    if (laborHours > 0) {
      return laborHours
    }
  } catch (error){
    console.warn(error)
  }
}

export function parseLanguages(repo) {
  const languages = get(repo, 'languages')
  if (Array.isArray(languages)) {
    return filter(languages, Boolean)
  } else {
    return []
  }
}

export function parseEmail(repo) {
  const email = get(repo, 'contact.email')
  if (Boolean(email)) {
    return email
  }
}

export function parseRepositoryURL(repo) {
  const url = get(repo, 'repositoryURL')
  if (Boolean(url)) {
    return url
  }
}

export function parseTags(repo) {
  try {
    return get(repo, 'tags').filter(Boolean)
  } catch (error) {
    return []
  }
}


export function getLastModifiedDateString(repo) {
  try {
    const dateLastModified = get(repo, 'date.lastModified') ||  get(repo, 'ghUpdatedAt')
    if (dateLastModified) {
      return new Date(dateLastModified).toLocaleDateString()
    }
  } catch (error) {
    console.warn(error)
  }
}
