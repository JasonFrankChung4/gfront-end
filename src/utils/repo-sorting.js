export function clean(string) {
  try {
    return string.toLowerCase().replace(/-/g,' ').trim()
  } catch (error) {
    return string
  }
}

export function getDate(repo, path) {
  if (repo.ghUpdatedAt) {
    return new Date(repo.ghUpdatedAt).getTime()
  } else if (repo.date && repo.date.lastModified) {
    return new Date(repo.date.lastModified).getTime()
  } else {
    return
  }
}

export function getTime(date) {
  try {
    if (date) {
      return new Date(date).getTime()
    } else {
      return -10e10
    }
  } catch (error) {
    return -10e10
  }
}

export function sortByBestMatch(a, b) {
  return Number(a.searchScore) > Number(b.searchScore) ? -1 : 1
}

export function sortByDataQuality(a, b) {
  return Number(a.score) > Number(b.score) ? -1 : 1
}

export function sortByDate(a, b) {
  try {
    const aDate = get(a, 'date.lastModified') || get(a, 'ghUpdatedAt')
    const bDate = get(b, 'date.lastModified') || get(b, 'ghUpdatedAt')
    const aTime = getTime(aDate)
    const bTime = getTime(bDate)
    return Math.sign(bTime - aTime) || 0;
  } catch (error) {
    return 0
  }
}

export function sortByName(a, b) {
  return clean(a.name) > clean(b.name) ? 1 : -1
}
