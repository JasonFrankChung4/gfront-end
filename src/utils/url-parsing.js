/* global URLSearchParams */

export function getParamAsArray(params, key) {
  if (params.has(key)) {
    return params.get(key)
      .split(',')
      .map(item => item.toLowerCase().trim())
  } else {
    return null
  }
}

export function getParamAsNumber(params, key) {
  if (params.has(key)) {
    return Number(params.get(key).toLowerCase().trim())
  } else {
    return null
  }
}

export function getParamAsString(params, key) {
  if (params.has(key)) {
    return params.get(key).toLowerCase().trim()
  } else {
    return null
  }
}

export function parseLocation(inpt) {
  const loc = inpt || window.location
  const pathname = loc.pathname
  const params = new URLSearchParams(loc.search)

  return {
    pathname,
    agencies: getParamAsArray(params, 'agencies'),
    languages: getParamAsArray(params, 'languages'),
    licenses: getParamAsArray(params, 'licenses'),
    skillLevels: getParamAsArray(params, 'skillLevels'),
    timeRequired: getParamAsArray(params, 'timeRequired'),
    page: getParamAsNumber(params, 'page'),
    sort: getParamAsString(params, 'sort'),
    query: getParamAsString(params, 'query')
  }
}
