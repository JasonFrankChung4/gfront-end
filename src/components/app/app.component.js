import React, { Component } from 'react'
import { Redirect } from 'react-router'
import { Route, Switch } from 'react-router-dom'
import { ConnectedRouter } from 'connected-react-router'
import history from 'browser-history'
import AboutPage from 'components/about-page'
import Roadmap from 'components/roadmap'
import Home from 'components/home'
import BrowseProjects from 'components/browse-projects'
import OpenTasks from 'components/open-tasks'
import ProjectPage from 'components/project-page'
import SearchPage from 'components/search-page'
import Menu from 'components/menu'
import Footer from 'components/footer'
import OfficialBanner from 'components/official-banner'
import PrivacyPolicy from 'components/privacy-policy'
import { refreshView } from 'utils/other'

export default class AppComponent extends Component {
  componentDidMount() {
    refreshView()
    window.addEventListener('popstate', _event => {
      this.props.rehydrate()
    })
  }

  render() {
    const location = window.location
    const isHomepage = location.pathname === '/'

    return (
      <ConnectedRouter history={history}>
        <div className="App">
          {isHomepage ? <OfficialBanner isDark /> : <OfficialBanner />}
          <Menu />
          <Switch location={location}>
            <Route exact path="/" component={Home} />
            <Route path="/search" component={SearchPage} />
            <Route path="/about" component={AboutPage} />
            <Route path="/browse-projects" component={BrowseProjects} />
            <Route path="/open-tasks" component={OpenTasks} />
            <Route path="/privacy-policy" component={PrivacyPolicy} />
            <Route path="/projects/:repoID" component={ProjectPage} />
            <Route path="/roadmap" component={Roadmap} />
            <Route path="/about" component={AboutPage} />
            <Redirect to="/" />
          </Switch>
          <Footer />
        </div>
      </ConnectedRouter>
    )
  }
}
