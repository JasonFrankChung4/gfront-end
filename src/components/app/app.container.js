import { forEach } from '@code.gov/cautious'
import updateBrowseParams from 'actions/update-browse-params'
import updateSearchParams from 'actions/update-search-params'
import updateTaskParams from 'actions/update-task-params'
import defaultState from 'constants/default-redux-store-state'
import { connect } from 'react-redux'
import { withRouter } from 'react-router-dom'
import { getConfigValue, now } from 'utils/other'
import { getNormalizedURLSearchParams, getSection } from 'utils/url-parsing'
import AppComponent from './app.component'

export const mapStateToProps = () => ({
  plugins: getConfigValue('plugins')
})

export const mapDispatchToProps = dispatch => ({
  rehydrate: () => {
    const section = getSection()
    if (section) {
      const parsed = getNormalizedURLSearchParams()
      if (section === 'browse' || section === 'search') {
        const filters = []
        forEach(['agencies', 'languages', 'licenses', 'usageTypes'], key => {
          console.log('key:', key)
          const values = parsed[key]
          if (values) {
            console.log('values:', values)
            values.forEach(value => {
              filters.push({ category: key, value, modified: now() })
            })
          }
        })
        const params = { filters }
        if (section === 'browse') {
          forEach(['page', 'sort', 'size'], key => {
            params[key] = parsed[key] || defaultState.browseParams[key]
          })
          dispatch(updateBrowseParams(params))
        } else if (section === 'search') {
          const params = { filters }
          forEach(['page', 'query', 'sort', 'size'], key => {
            params[key] = parsed[key] || defaultState.searchParams[key]
          })
          dispatch(updateSearchParams(params))
        }
      } else if (section === 'tasks') {
        const filters = []
        forEach(['agencies', 'languages', 'skillLevels', 'timeRequired', 'usageTypes'], key => {
          const values = parsed[key]
          if (values) {
            values.forEach(value => {
              filters.push({ category: key, value, modified: now() })
            })
          }
        })
        const params = { filters }
        forEach(['page', 'size'], key => {
          params[key] = parsed[key] || defaultState.taskParams[key]
        })
        dispatch(updateTaskParams(params))
      }
    }
  }
})

const AppContainer = connect(
  mapStateToProps,
  mapDispatchToProps
)(AppComponent)

export default withRouter(AppContainer)
