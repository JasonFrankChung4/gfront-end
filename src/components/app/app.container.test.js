import pick from 'lodash.pick'

import updateBrowseParams from 'actions/update-browse-params'
import updateSearchParams from 'actions/update-search-params'
import updateTaskParams from 'actions/update-task-params'
import { getNormalizedURLSearchParams, getSection } from 'utils/url-parsing'
import { now } from 'utils/other'
import defaultState from 'constants/default-redux-store-state'
import { mapStateToProps, mapDispatchToProps } from 'components/app/app.container'

jest.mock('actions/update-browse-params')
jest.mock('actions/update-search-params')
jest.mock('actions/update-task-params')
jest.mock('utils/url-parsing')

const searchParams = {
  agencies: ['a1', 'a2'],
  languages: ['l1'],
  page: 'test-page',
  sort: 'test-sort',
  size: 'test-size',
  query: 'test-query',
}

// filters that get mapped out from `searchParams`
const filters =  [
  { category: 'agencies', value: 'a1', modified: now() },
  { category: 'agencies', value: 'a2', modified: now() },
  { category: 'languages', value: 'l1', modified: now() },
]

const dispatch = jest.fn()

// helper function for testing `rehydrate` dispatches an action with expected params
const testBySection = ({ section, action, expected }) => {
  getSection.mockImplementation(() => section)
  mapDispatchToProps(dispatch).rehydrate()
  expect(dispatch).toBeCalled()
  expect(action).toBeCalledWith(expected)
}

describe('containers - App', () => {
  beforeEach(() => {
    getNormalizedURLSearchParams.mockImplementation(() => searchParams)
  })

  describe('mapStateToProps', () => {
    it('should return the correct properties', () => {
      expect(mapStateToProps()).toMatchSnapshot()
    })
  })

  describe('mapDispatchToProps', () => {
    describe('rehydrate', () => {
      describe('section is `browse`', () => {
        it('should dispatch the `updateBrowseParams` action with the correct params', () => {
          const expected = { ...pick(searchParams, ['page', 'sort', 'size']), filters }
          testBySection({ section: 'browse',  action: updateBrowseParams, expected })
        })

        it('should dispatch the `updateBrowseParams` action with the default browse params if none provided', () => {
          const expected = pick(defaultState.browseParams, ['page', 'sort', 'size', 'filters'])
          getNormalizedURLSearchParams.mockImplementation(() => ({}))
          testBySection({ section: 'browse',  action: updateBrowseParams, expected })
        })
      })
    })

    describe('section is `search`', () => {
      it('should dispatch the `updateSearchParams` action with the correct params', () => {
        const expected = { ...pick(searchParams, ['page', 'query', 'sort', 'size']), filters }
        testBySection({ section: 'search',  action: updateSearchParams, expected })
      })

      it('should dispatch the `updateSearchParams` action with the default search params if none provided', () => {
        const expected = pick(defaultState.searchParams, ['page', 'query', 'sort', 'size', 'filters'])
        getNormalizedURLSearchParams.mockImplementation(() => ({}))
        testBySection({ section: 'search',  action: updateSearchParams, expected })
      })
    })

    describe('section is `tasks`', () => {
      it('should dispatch the `updateTaskParams` action with the correct params if the section is `tasks`', () => {
        const expected = { ...pick(searchParams, ['page', 'size']), filters }
        testBySection({ section: 'tasks',  action: updateTaskParams, expected })
      })

      it('should dispatch the `updateTaskParams` action with the default task params if none provided', () => {
        const expected = pick(defaultState.taskParams, ['page', 'query', 'sort', 'size', 'filters'])
        getNormalizedURLSearchParams.mockImplementation(() => ({}))
        testBySection({ section: 'tasks',  action: updateTaskParams, expected })
      })
    })

    describe('section is unknown or undefined', () => {
      it('should not dispatch any actions', () => {
        getSection.mockImplementation(() => 'unknown')
        mapDispatchToProps(dispatch).rehydrate()
        expect(dispatch).not.toBeCalled()

        getSection.mockImplementation(() => undefined)
        mapDispatchToProps(dispatch).rehydrate()
        expect(dispatch).not.toBeCalled()
      })
    })
  })
})
