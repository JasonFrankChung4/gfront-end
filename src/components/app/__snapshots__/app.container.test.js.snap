// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers - App mapStateToProps should return the correct properties 1`] = `
Object {
  "plugins": Array [
    Object {
      "component": "@code.gov/about-page",
      "filename": "about-page.js",
      "route": "/about",
    },
  ],
}
`
